tests/rptest/tests/e2e_shadow_indexing_test.py

```diff
@@ -445,4 +445,177 @@ def test_consumer_groups(self):
+        self.logger.debug(
+            f'node: {node.name}, path: {topic_path}, partition dirs: {partition_dirs}'
+        )
+
+        for part_dir in partition_dirs:
+            part_path = f'{topic_path}/{part_dir}'
+            files = self._listdir(part_path, node)
+
+            self.logger.debug(
+                f'node: {node.name}, path: {part_path}, files: {files}')
+            for _file in files:
+                self.logger.debug(f'file: {_file}')
+                # The regex is from src/v/storage/fs_utils.h::parse_segment_filename
+                match = re.match(r'(^(\d+)-(\d+)-([\x00-\x7F]+).log$)', _file)
+                if match:
+                    segments.append(_file)
+
+    def _find_segment_files_on_all_nodes(self, segments):
+        topic_path = f'/var/lib/redpanda/data/kafka/{self.topic}'
+
+        for node in self.redpanda.nodes:
+            self._find_segment_files_on_node(topic_path, node, segments)
+
+    def _check_si_cache_before_fetch(self):
+        # SI cache should be empty on all nodes
+        for node in self.redpanda.nodes:
+            contents = self._listdir(
+                '/var/lib/redpanda/data/cloud_storage_cache', node)
+            assert contents == [], f'SI cache on {node.name} is not empty before fetch'
+
+    def _check_si_cache_after_fetch(self, data_dir_segments):
+        # Sample file path:
+        # /var/lib/redpanda/data/cloud_storage_cache/<some string>/kafka/<topic name>/<partition>/<segment file>
+        root = '/var/lib/redpanda/data/cloud_storage_cache'
+
+        segment_files = []
+
+        # Get the paths to the partition dirs within the
+        # SI cache
+        for node in self.redpanda.nodes:
+            all_parts_dirs = []
+            some_dirs = self._listdir(root, node, only_dirs=True)
+
+            for d in some_dirs:
+                path = f'{root}/{d}/kafka/{self.topic}'
+                parts_dirs = self._listdir(path, node, only_dirs=True)
+
+                # Merge the lists of partitions together.
+                all_parts_dirs += parts_dirs
+
+            for part_dir in all_parts_dirs:
+                path = f'{root}/{d}/kafka/{self.topic}/{part_dir}'
+                files = self._listdir(path, node)
+
+                # The data within the SI cache is a .log.N segment file
+                # where N is an positive integer
+                # Trim off the .N of the segment file
+                temp = [_file[:-2] for _file in files]
+
+                # Merge the list of files
+                segment_files += temp
+
+        for segment_file in segment_files:
+            match = re.match(r'(^(\d+)-(\d+)-([\x00-\x7F]+).log$)',
+                             segment_file)
+            assert match, f'Invalid segment filename {segment_file}'
+            assert segment_file in data_dir_segments, f'{segment_file} was not in data dir'
+
+    @cluster(num_nodes=5)
+    def test_data_loss(self):
```

Comment:
I think in test planning when we talked about "data loss" we might have meant deleting objects from S3 and checking that shadow indexing handles that gracefully (@Lazin do you remember?)

This test of segments being removed on disk is useful, but could probably be combined with another -- anywhere we do reads from shadow indexing and see items in the cache, we can check those cached segments are not also present in the data directory, and thereby conclude that the cleanup was happening properly.

Reply:
> I think in test planning when we talked about "data loss" we might have meant deleting objects from S3 and checking that shadow indexing handles that gracefully (@Lazin do you remember?)

This test was specified to be this way in the testing strategy gdoc we were editing. I tagged you in the relevant comment in the gdoc.

> This test of segments being removed on disk is useful, but could probably be combined with another -- anywhere we do reads from shadow indexing and see items in the cache, we can check those cached segments are not also present in the data directory, and thereby conclude that the cleanup was happening properly.

Sure I could extend the redpanda service or some utility class to have a `check_si_cache(callback)` where callback is invoked on every segment filename.

Reply:
I stand corrected!  Let's tweak the naming of the test a bit, I think "data loss" is understood by most folks to mean unexpected data loss, rather than routine cleanup.